Add sorting to data table

AmigoBongero opened the pull request:
- Dev: developed sorting (js logic and markdown) for every column in tabs in accountsComponent and expensesTabComponent. 

Time Spent: 4 hours.

AmigoBongero committed:
Added sorting in datatables

- dev: sorting logic and markdown in accounts and expenses tabs
- format: changed fieldName in account columns constant

AmigoBongero committed:
Merge branch 'Implement-infinite-scrolling-in-Account-component' into Add-sorting-to-data-table

# Conflicts:
#	DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.html
#	DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js

AmigoBongero committed:
Space

AmigoBongero committed:
Added component to lightning tab

AmigoBongero committed:
Fix

IJWMG commented on DevJourney/force-app/main/default/classes/AccountsComponentController.cls:
```diff
@@ -1,4 +1,4 @@
-public with sharing class AccountsComponentController {  
+public with sharing class AccountsComponentController {
```
@AmigoBongero Won't this changes conflict with other PRs?

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.css:
@AmigoBongero We don't need this one here

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -4,55 +4,131 @@ import { ShowToastEvent } from 'lightning/platformShowToastEvent';
+    recentlyViewedAccountsFullData = [];
+    recentlyViewedAccountRecordCount = 20;
+    recentlyViewedAccountsSortDirection = 'asc';
+    recentlyViewedAccountsSortedBy;
```
@AmigoBongero Assign default value

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -4,55 +4,131 @@ import { ShowToastEvent } from 'lightning/platformShowToastEvent';
+        let fieldName = field;
+        let dataToSort = [...data];
+
+        let keyValue = (a) => {
```
@AmigoBongero Variable naming, also do you need a function here?

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -4,55 +4,131 @@ import { ShowToastEvent } from 'lightning/platformShowToastEvent';
+            return a[fieldName];
+        };
+
+        let isReverse = direction === 'asc' ? 1: -1;
```
@AmigoBongero Whitespace

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js-meta.xml:
```diff
@@ -4,5 +4,6 @@
     <isExposed>true</isExposed>
     <targets>
         <target>lightning__AppPage</target>
+        <target>lightning__Tab</target>
```
@AmigoBongero Is this for remving a line in the top of the page?

AmigoBongero replied:
It is)

IJWMG reviewed (changes requested):
Ok. I will review it ater you merge prev PR into main and back here. Let me now whent it will be ready

AmigoBongero committed:
Merge branch 'main' into Add-sorting-to-data-table

# Conflicts:
#	DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js

AmigoBongero committed:
Fixed merge mistakes

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -42,6 +52,18 @@ export default class AccountsComponent extends LightningElement {
+    handleMyAccountsSort(event) {
+        this.myAccountsSortedBy = event.detail.fieldName;
+        this.myAccountsSortDirection = event.detail.sortDirection;
+        this.myAccountsData = this.sortData(this.myAccountsData, this.myAccountsSortedBy, this.myAccountsSortDirection);
```
@AmigoBongero So what if I sort it out by name desc?

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -42,6 +52,18 @@ export default class AccountsComponent extends LightningElement {
+    handleRecentlyViewedAccountsSort(event) {
+        this.recentlyViewedAccountsSortedBy = event.detail.fieldName;
+        this.recentlyViewedAccountsSortDirection = event.detail.sortDirection;
+        this.recentlyViewedAccountsData = this.sortData(this.recentlyViewedAccountsData, this.recentlyViewedAccountsSortedBy, this.recentlyViewedAccountsSortDirection);
```
@AmigoBongero New question, What if I sort it by Phone and then scroll down?

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -59,6 +81,24 @@ export default class AccountsComponent extends LightningElement {
+        let isReverse = direction === 'asc' ? 1: -1;
+
+        dataToSort.sort((x, y) => {
+            x = keyValue(x) ? keyValue(x) : '';
```
@AmigoBongero First of all, `x = keyValye(x) ?? '';`. Second of all, name the variables properly

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -87,6 +127,7 @@ export default class AccountsComponent extends LightningElement {
             });
     }
 
+
```
@AmigoBongero remove one empty line

IJWMG commented on DevJourney/force-app/main/default/lwc/expensesTabComponent/expensesTabComponent.js:
```diff
@@ -28,6 +28,8 @@ export default class ExpensesTabComponent extends LightningElement {
     expensesFullData = [];
     expensesRecordCount = 20;
+    sortDirection = 'asc';
+    sortedBy;
```
@AmigoBongero Initialize variable

IJWMG commented on DevJourney/force-app/main/default/lwc/expensesTabComponent/expensesTabComponent.js:
```diff
@@ -140,6 +148,24 @@ export default class ExpensesTabComponent extends LightningElement {
      * @description     Reusable Code.
      */
+    sortData(fieldName, direction) {
+        let parseData = JSON.parse(JSON.stringify(this.expensesData));
```
@AmigoBongero Is it necessary here to make this kind of copy? Also what if we sort it by any field other than name and then scroll down?

IJWMG commented on DevJourney/force-app/main/default/lwc/expensesTabComponent/expensesTabComponent.js:
```diff
@@ -140,6 +148,24 @@ export default class ExpensesTabComponent extends LightningElement {
      */
+    sortData(fieldName, direction) {
+        let parseData = JSON.parse(JSON.stringify(this.expensesData));
+        // Return the value stored in the field
```
@AmigoBongero Apply comments from accountsComponent.js

AmigoBongero committed:
Logic improvements  + corrections

- dev: rewrote sorting logic
- format: made clearer variable namings
- format: whitespace corrections + variables initialization

AmigoBongero reviewed (commented):
Rewrote sorting logic + clearer variable namings  + code formatting

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -59,6 +83,30 @@ export default class AccountsComponent extends LightningElement {
+        let getFieldValue = (record) => {
+            return record[fieldName] ? record[fieldName].toString().toLowerCase() : '';
+        };
+        let isReverse = direction === 'asc' ? 1 : -1;
```
@AmigoBongero This is not a boolean variable

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -4,21 +4,31 @@ import { ShowToastEvent } from 'lightning/platformShowToastEvent';
     myAccountsData = [];
     myAccountsFullData = [];
     myAccountsRecordCount = 20;
+    myAccountsSortDirection = 'asc';
```
@AmigoBongero Do you ever use this variables outside of corresponding event context?

AmigoBongero replied:
They are used in markdown.

IJWMG commented on DevJourney/force-app/main/default/lwc/accountsComponent/accountsComponent.js:
```diff
@@ -59,6 +83,30 @@ export default class AccountsComponent extends LightningElement {
+        };
+        let isReverse = direction === 'asc' ? 1 : -1;
+
+        sortedData.sort((x, y) => {
```
@AmigoBongero Why are you stuck with x and y? I think it is crearer to use firtsValue and secondValue, or something like this.

IJWMG commented on DevJourney/force-app/main/default/lwc/expensesTabComponent/expensesTabComponent.js:
```diff
@@ -140,6 +148,35 @@ export default class ExpensesTabComponent extends LightningElement {
     /*
      * @description     Reusable Code.
      */
+    sortData(fieldName, direction) {
```
@AmigoBongero Wow, such a huge amount of code. Seems generic (except for the last two lines and anount logic). That's good. If only there was a way to use it from one place everywhere we need... There is also a better way to compare values for sort :)

AmigoBongero committed:
Improvements

AmigoBongero reviewed (commented):
Impovements + fixes + utility component

IJWMG commented on DevJourney/force-app/main/default/lwc/utilityComponent/utilityComponent.js:
```diff
@@ -0,0 +1,27 @@
+export function sortData(data, fieldName, direction) {
+      let sortedData = [...data];
+      let getFieldValue = (record) => {
```
@AmigoBongero So, what if we are sorting something without field value?

AmigoBongero replied:
So in 'return' we return an empty string if the field value is not present. This ensures that the logic sorts records without field values correctly. Is this wrong?

IJWMG replied:
@AmigoBongero So let's name it something like 'sortArrayOfObjectsByField'. Because now I want to sort everything with this function :)

IJWMG commented on DevJourney/force-app/main/default/lwc/utilityComponent/utilityComponent.js:
```diff
@@ -0,0 +1,27 @@
+      let getFieldValue = (record) => {
+          return record[fieldName] ? record[fieldName].toString().toLowerCase() : '';
+      };
+      let getFieldDataType = (value) => {
```
@AmigoBongero What is this?

AmigoBongero replied:
The 'Amount' field haven't been sorted correctly. As I understood it was sorted as string, so I got wrong sorting. I decided to add this check to ensure the field data type. And it worked.

IJWMG replied:
@AmigoBongero Sooooo, why don't you just `if (isFinite(firstFieldValue) && isFinite(secondFieldValue)) { \n //do stuff`?

AmigoBongero replied:
Great idea)

IJWMG commented on DevJourney/force-app/main/default/lwc/utilityComponent/utilityComponent.js:
```diff
@@ -0,0 +1,27 @@
+          return record[fieldName] ? record[fieldName].toString().toLowerCase() : '';
+      };
+      let getFieldDataType = (value) => {
+          if (!isNaN(value) && isFinite(value)) return 'integer';
```
@AmigoBongero Return on the next line, use"{}". Also '!isNan()' here does nothing. isFinite() checks for the NaN value as well

IJWMG commented on DevJourney/force-app/main/default/lwc/utilityComponent/utilityComponent.js:
```diff
@@ -0,0 +1,27 @@
+      };
+      let sortOrder = direction === 'asc' ? 1 : -1;
+
+      sortedData.sort((firstValue, secondValue) => {
```
@AmigoBongero There is still a better way for this, but if it works its fine

IJWMG commented on DevJourney/force-app/main/default/lwc/utilityComponent/utilityComponent.js:
```diff
@@ -0,0 +1,27 @@
+      let sortOrder = direction === 'asc' ? 1 : -1;
+
+      sortedData.sort((firstValue, secondValue) => {
+          let xValue = getFieldValue(firstValue);
```
@AmigoBongero Still x and y...

IJWMG commented on DevJourney/force-app/main/default/lwc/utilityComponent/utilityComponent.js:
```diff
@@ -0,0 +1,27 @@
+          let xFieldDataType = getFieldDataType(xValue);
+          let yFieldDataType = getFieldDataType(yValue);
+
+          if (xFieldDataType === 'integer' && yFieldDataType === 'integer') {
```
@AmigoBongero So, if Amount is 5.64$ we are going to compare it as a string?

AmigoBongero committed:
Utility component impovements

AmigoBongero reviewed (commented):
utilityComponent fixes

AmigoBongero committed:
sortData logic and naming improvements

AmigoBongero reviewed (commented):
sortData logic + naming improvements

IJWMG commented on DevJourney/force-app/main/default/lwc/utilityComponent/utilityComponent.js:
```diff
@@ -0,0 +1,28 @@
+          let secondFieldValue = getFieldValue(secondRecord);
+
+          if (isFinite(firstFieldValue) && isFinite(secondFieldValue)) {
+              firstFieldValue = firstFieldValue === '' ? null : parseInt(firstFieldValue);
```
@AmigoBongero So, 5.25$, 5.27$ and 5.89$ in the Amount field. How will it be sorted?

AmigoBongero committed:
Returned parseFloat
